### PR TITLE
Adding test when registering a batching rule for a CompositeImplicitAutograd operation

### PR DESCRIPTION
**Pull request by srossross**
This is a Follow on from https://github.com/pytorch/pytorch/pull/88771 which should close out https://github.com/pytorch/functorch/issues/1009 I've got another PR where I'm moving some operators over https://github.com/pytorch/pytorch/pull/89762

you can see that the new test file is being picked [run here](https://github.com/pytorch/pytorch/actions/runs/3617298059/jobs/6096218583#step:10:472)

### CODE REVIEW

File: test/functorch/test_dispatch.py [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+    ):
+        assert (
+            registration not in FuncTorchBatchedRegistrations
+        ), f"{registration} is also registered in FuncTorchBatched"
+
+    @dispatch_registrations("FuncTorchBatchedDecomposition", xfail_functorch_batched_decomposition)
+    def test_register_functorch_batched_decomposition(self, registration):
+        assert (
+            registration in CompositeImplicitAutogradRegistrations
+        ), f"{registration} should also be CompositeImplicitAutograd"
```
Comment (srossross) [inline test/functorch/test_dispatch.py:129]: I'm not sure if this is a good test as well

File: test/functorch/test_dispatch.py [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+
+xfail_functorch_batched_decomposition = {
+    'aten::diagonal_copy',
+    'aten::is_same_size',
+    'aten::t',
+    'aten::t_',
+    'aten::unfold_copy',
+}
+
+class dispatch_registrations(_TestParametrizer):
```
Comment (srossross) [inline test/functorch/test_dispatch.py:90]: I kept this class in the test file but happy to move it out if it makse sense.

**Comment by zou3519** [comment]
> you can see that the new test file is being picked [run here](https://github.com/pytorch/pytorch/actions/runs/3617298059/jobs/6096218583#step:10:472)

Always good to check, thanks for checking

File: test/functorch/test_dispatch.py [hunk 3]
```diff
@@ -0,0 +1,115 @@
+# Owner(s): ["module: functorch"]
+from functools import wraps
+import unittest
+
```
Comment (zou3519) [inline test/functorch/test_dispatch.py:4]: nit: test_vmap_registrations is a better name for this file.

Also is there a reason why you didn't put all of this into functorch/test_vmap? (I'm neutral either way)

File: test/functorch/test_dispatch.py (hunk 2 above)
Comment (zou3519) [reply test/functorch/test_dispatch.py:70]: Here seems fine for now, if someone else needs it they can move it out

File: test/functorch/test_dispatch.py [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+
+class TestFunctorchDispatcher(TestCase):
+    @dispatch_registrations("CompositeImplicitAutograd", xfail_functorch_batched)
+    def test_register_a_batching_rule_for_composite_implicit_autograd(
+        self, registration
+    ):
+        assert (
+            registration not in FuncTorchBatchedRegistrations
+        ), f"{registration} is also registered in FuncTorchBatched"
```
Comment (zou3519) [inline test/functorch/test_dispatch.py:103]: Let's be more prescriptive with the error messages. If I had no experience working with functorch and saw this error message I'd be confused.

"You've added a batching rule for a CompositeImplicitAutograd operator {registration}. The correct way to add vmap support for it is to put it into BatchRulesDecomposition to reuse the CompositeImplicitAutograd decomposition".

**Review by zou3519 — APPROVED** [review]
LGTM. I had some comments about making the error message more prescriptive (because the people who run into this are usually not the people who directly develop functorch), please read

File: test/functorch/test_dispatch.py (hunk 3 above)
Comment (srossross) [reply test/functorch/test_dispatch.py:4]: > nit: test_vmap_registrations is a better name for this file.

👍 

No, I was going off the the similar `test_find_dangling_impls` which was in test/test_dispatch.py

File: test/functorch/test_dispatch.py [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+
+    @dispatch_registrations("FuncTorchBatchedDecomposition", xfail_functorch_batched_decomposition)
+    def test_register_functorch_batched_decomposition(self, registration):
+        assert (
+            registration in CompositeImplicitAutogradRegistrations
+        ), f"{registration} should also be CompositeImplicitAutograd"
```
[the remark this answers is not in the document]
Comment (srossross) [reply test/functorch/test_dispatch.py]: done in a032b7a3845a68d2429c8613ff05054b3d53bff4

File: test/functorch/test_dispatch.py (hunk 4 above)
Comment (srossross) [reply test/functorch/test_dispatch.py]: done in a032b7a3845a68d2429c8613ff05054b3d53bff4

File: test/functorch/test_dispatch.py (hunk 3 above)
Comment (srossross) [reply test/functorch/test_dispatch.py]: done in https://github.com/pytorch/pytorch/pull/89465/commits/249ace286845b208455bc8b73cde7e84d49bc852

**Pull request closed without merging**